src/server/pfs/server/source.go

```diff
@@ -0,0 +1,255 @@
+	}
+	return &FileReader{
+		file: file,
+		idx:  idx,
+		fmr:  fmr,
+		mr:   mr,
+		hash: h,
+	}
+}
+
+func (fr *FileReader) updateFileInfo(idx *index.Index) {
+	fr.fileCount++
+	for _, dataRef := range idx.DataOp.DataRefs {
+		fr.hash.Write([]byte(dataRef.Hash))
+	}
+}
+
+// Info returns the info for the file.
+func (fr *FileReader) Info() *pfs.FileInfoV2 {
+	return &pfs.FileInfoV2{
+		File: fr.file,
+		Hash: pfs.EncodeHash(fr.hash.Sum(nil)),
+	}
+}
+
+// Get writes a tar stream that contains the file.
+func (fr *FileReader) Get(w io.Writer, noPadding ...bool) error {
+	if err := fr.fmr.Get(w); err != nil {
+		return err
+	}
+	for fr.fileCount > 0 {
+		fmr, err := fr.mr.Next()
+		if err != nil {
+			return err
+		}
+		if err := fmr.Get(w); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	if len(noPadding) > 0 && noPadding[0] {
+		return nil
+	}
+	// Close a tar writer to create tar EOF padding.
+	return tar.NewWriter(w).Close()
+}
+
+func (fr *FileReader) drain() error {
+	for fr.fileCount > 0 {
+		if _, err := fr.mr.Next(); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	return nil
+}
+
+// Source iterates over FileInfoV2s generated from a fileset.Source
+type Source struct {
+	commit        *pfs.Commit
+	getReader     func() fileset.FileSource
+	computeHashes bool
+}
+
+// Creates a Reader which emits FileInfoV2s with the information from commit, and the entries from readers
+// returned by getReader.  If getReader returns different Readers all bets are off.
+func NewSource(commit *pfs.Commit, computeHashes bool, getReader func() fileset.FileSource) *Source {
+	return &Source{
+		commit:        commit,
+		getReader:     getReader,
+		computeHashes: computeHashes,
+	}
+}
+
+func (s *Source) Iterate(ctx context.Context, cb func(*pfs.FileInfoV2, fileset.File) error) error {
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	fs1 := s.getReader()
+	fs2 := s.getReader()
+	s2 := newStream(ctx, fs2)
+	cache := make(map[string][]byte)
+	return fs1.Iterate(ctx, func(fr fileset.File) error {
+		idx := fr.Index()
+		fi := &pfs.FileInfoV2{
+			File: client.NewFile(s.commit.Repo.Name, s.commit.ID, idx.Path),
+		}
+		if s.computeHashes {
+			var err error
+			var hashBytes []byte
+			if indexIsDir(idx) {
+				hashBytes, err = computeHash(cache, s2, idx.Path)
+				if err != nil {
+					return err
+				}
+			} else {
+				hashBytes = computeFileHash(idx)
+			}
+			fi.Hash = pfs.EncodeHash(hashBytes)
+		}
+		if err := cb(fi, fr); err != nil {
+			return err
+		}
+		delete(cache, idx.Path)
+		return nil
+	})
+}
+
+func computeHash(cache map[string][]byte, s *stream, x string) ([]byte, error) {
+	if hashBytes, exists := cache[x]; exists {
+		return hashBytes, nil
+	}
+	// check that the next item in the stream is x
+	fr, err := s.Peek()
+	if err != nil {
+		if err == io.EOF {
+			return nil, errors.Errorf("stream is done, can't compute hash for %s", x)
+		}
+		return nil, err
+	}
+	idx := fr.Index()
+	if idx.Path != x {
```

Comment:
Not a big fan of the use of the variable name `x` here and in other places. Even though it is locally scoped, nothing about the letter `x` is relevant to the path we are computing the hash of. I think some options for better names would be `pathStr` or `idxPath`. In general lets try to have the variable names be meaningful, even the locally scoped ones. I'm not against single letter local variables when the purpose of the variable is clear. This is just a string with the variable name `x` which tells me nothing about the meaning of this variable in this function.

Reply:
I'll change it here.  For simple pure functions, like `cleanPath` for example, there is no better argument name than `in` or `input` or `x`, which has a pretty deep tradition of being used as an input variable in mathematics. It's also one character.

Reply:
I mean `pathStr` would be a pretty clear name for the input to a `cleanPath` function. Conventions for expressing mathematical functions are not always clear when trying to apply them to programming. I don't really care too much about the `cleanPath` case though because that is simple enough to intuit the purpose of the variable.